app/models/grouping.rb

```diff
@@ -445,12 +446,17 @@ def select_starter_file_entries
       assignment.default_starter_file_group&.starter_file_entries || []
     when 'shuffle'
       assignment.starter_file_groups.includes(:starter_file_entries).map do |g|
-        StarterFileEntry.find_by(id: g.starter_file_entries.ids.sample)
+        # If this grouping has previous starter files, try to choose an entry with the same path as before
+        old_entry = g.starter_file_entries.where(path: self.starter_file_entries.map(&:path)).sample
```

Comment:
The use of `sample` here is pretty non-obvious. I think I know what you're trying to do but it would be great to leave a more explanatory comment explaining the use. (Personally I would have just used a `find_by`, but you can leave the `sample` if you want.)